sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoder.java

```diff
@@ -123,28 +124,31 @@ public void verifyDeterministic()
 
   private void verifyDeterministic(Schema schema)
       throws org.apache.beam.sdk.coders.Coder.NonDeterministicException {
```

Comment:
Thanks for the catch! Would it work to replace this with (the simpler):

Iterable<Coder<?>> coders = schema.getFields().stream()
        .map(Field::getType).map(RowCoder::coderForFieldType).collect(Collectors.toList());
Coder.verifyDeterministic(this, "All fields must have deterministic encoding", coders);

Your error message is more descriptive, but maybe better to make the error-message improvements to Coder.verifyDeterministic. That way all coders benefit, not just RowCoder. These improvements are separable too - can be done in a separate PR.

Reply:
I've started looking into `RowCoder#verifyDeterministic` when I was debugging pipeline that was failing due non-deterministic exception. It was hard to figure out what is wrong without having a stack trace, and detailed error message saying which field is causing the problem. 

I didn't find two things:
- stack trace to the first coder violating determinism
- field name so that I can understand what is failing
- in case of SchemaCoder, I would propose to add a class name

There is a similar code in `StructuredCoder`, but it doesn't have field names. 

I agree to go with the most simple approach and improve error messages later.

Reply:
Thanks. Improving error messages in Coder itself will help all Beam pipelines, not just the ones using schemas.